File: algorithms/graph/chromatic-number/typescript/chromatic-number.ts
export function chromaticNumber(arr: number[]): number {
  if (arr.length < 2) return 0;
  const n = arr[0];
  const m = arr[1];

  if (arr.length < 2 + 2 * m) return 0;
  if (n === 0) return 0;

  const adj: boolean[][] = Array.from({ length: n }, () => Array(n).fill(false));
  for (let i = 0; i < m; i++) {
    const u = arr[2 + 2 * i];
    const v = arr[2 + 2 * i + 1];
    if (u >= 0 && u < n && v >= 0 && v < n) {
      adj[u][v] = true;
      adj[v][u] = true;
    }
  }

  const color: number[] = new Array(n).fill(0);

  function isSafe(u: number, c: number): boolean {
    for (let v = 0; v < n; v++) {
      if (adj[u][v] && color[v] === c) {
        return false;
      }
    }
    return true;
  }

  function graphColoringUtil(u: number, k: number): boolean {
    if (u === n) return true;

    for (let c = 1; c <= k; c++) {
      if (isSafe(u, c)) {
        color[u] = c;
        if (graphColoringUtil(u + 1, k)) {
          return true;
        }
        color[u] = 0;
      }
    }
    return false;
  }

  for (let k = 1; k <= n; k++) {
    if (graphColoringUtil(0, k)) {
      return k;
    }
  }

  return n;
}

File: algorithms/graph/chromatic-number/rust/chromatic_number.rs
pub fn chromatic_number(arr: &[i32]) -> i32 {
    if arr.len() < 2 {
        return 0;
    }
    let n = arr[0] as usize;
    let m = arr[1] as usize;

    if arr.len() < 2 + 2 * m {
        return 0;
    }
    if n == 0 {
        return 0;
    }

    let mut adj = vec![vec![false; n]; n];
    for i in 0..m {
        let u = arr[2 + 2 * i] as usize;
        let v = arr[2 + 2 * i + 1] as usize;
        if u < n && v < n {
            adj[u][v] = true;
            adj[v][u] = true;
        }
    }

    let mut color = vec![0; n];

    for k in 1..=n {
        if graph_coloring_util(0, n, k as i32, &mut color, &adj) {
            return k as i32;
        }
    }

    n as i32
}

fn is_safe(u: usize, c: i32, n: usize, color: &[i32], adj: &Vec<Vec<bool>>) -> bool {
    for v in 0..n {
        if adj[u][v] && color[v] == c {
            return false;
        }
    }
    true
}

fn graph_coloring_util(u: usize, n: usize, k: i32, color: &mut Vec<i32>, adj: &Vec<Vec<bool>>) -> bool {
    if u == n {
        return true;
    }

    for c in 1..=k {
        if is_safe(u, c, n, color, adj) {
            color[u] = c;
            if graph_coloring_util(u + 1, n, k, color, adj) {
                return true;
            }
            color[u] = 0;
        }
    }
    false
}
